xla/backends/gpu/runtime/command_buffer_cmd.cc

```diff
@@ -1449,22 +1468,73 @@ absl::StatusOr<const se::CommandBuffer::Command*> WhileCmd::Record(
-            CreateCommands(&cond_commands_, &execute_params, &record_params),
-            CreateCommands(&body_commands_, &execute_params, &record_params),
-            dependencies);
-      },
-      [&](const se::CommandBuffer::Command* command) {
-        return command_buffer->UpdateWhile(
-            command, se::DeviceMemory<bool>(pred),
-            UpdateCommands(&cond_commands_, &execute_params, &record_params),
-            UpdateCommands(&body_commands_, &execute_params, &record_params));
-      });
+  if (is_unrolled_loop_) {
+    // When the loop is unrolled, we need to record the body commands for
+    // `trip_count` times into child_command_buffer_, and implement the While
+    // command as a child command.
+    VLOG(3) << "Recording unrolled loop with trip_count: "
+            << trip_count_.value();
+    CHECK(child_command_buffer_ != nullptr);
+
+    // Unroll the while loop body for `trip_count` times.
+    // Unrolled execution sequence: cond -> body -> cond -> body -> ...
+    // In the unrolled pattern, we still need to run the cond commands because
+    // body commands might depends on the value of index variable that is
+    // updated by condition commands.
+    auto new_record_params = record_params;
+    for (int64_t i = 0; i < trip_count_.value(); ++i) {
+      new_record_params.unroll_iteration = i;
+      if (i == 0) {
+        // First iteration, cond_commands_ will not have dependencies.
+        TF_RETURN_IF_ERROR(cond_commands_.Record(
+            execute_params, new_record_params, CommandBufferCmd::RecordCreate{},
+            child_command_buffer_.get(), false));
+      } else {
+        // Other iterations, cond_commands_ will have dependencies on the sink
+        // commands from previous iteration's body.
+        auto body_sink_commands = body_commands_.SinkCommands(
+            new_record_params, child_command_buffer_.get(), i - 1);
+        TF_RETURN_IF_ERROR(cond_commands_.Record(
+            execute_params, new_record_params,
+            CommandBufferCmd::RecordCreate{absl::MakeSpan(body_sink_commands)},
+            child_command_buffer_.get(), false));
+      }
+      auto cond_sink_commands = cond_commands_.SinkCommands(
+          new_record_params, child_command_buffer_.get(), i);
+      TF_RETURN_IF_ERROR(body_commands_.Record(
+          execute_params, new_record_params,
+          CommandBufferCmd::RecordCreate{absl::MakeSpan(cond_sink_commands)},
+          child_command_buffer_.get(),
+          i == trip_count_.value() - 1 ? true : false));
+    }
+    return Handle(
+        std::move(record_action),
+        [&](absl::Span<const se::CommandBuffer::Command* const> dependencies) {
+          return command_buffer->CreateChildCommand(
+              se::CommandBuffer::ChildCommandType::kMoved,
+              *child_command_buffer_, dependencies);
+        },
+        [&](const se::CommandBuffer::Command* command) {
+          // Moved child command does not need to be updated.
```

Comment:
Hm, I don't think it's true, if buffer allocation changes you need to update all nodes that were created from an unrolled loop?

Reply:
Update the comments to describe the update process more precisly.